// react-here-maps/dist/mixins/h-map-methods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var ReactDOM = require("react-dom");
var mixin_1 = require("../utils/mixin");
exports.HMapMethods = mixin_1.default({
    // return the HTMLElement representing this HEREMap component
    getElement() {
        return ReactDOM.findDOMNode(this);
    },
    getMap() {
        return this.state.map;
    },
    setCenter(point) {
        const { animateCenter } = this.props;
        const { map } = this.state;
        map.setCenter(point, animateCenter === true);
    },
    setZoom(zoom) {
        const { animateZoom } = this.props;
        const { map } = this.state;
        map.setZoom(zoom, animateZoom === true);
    },
    // change the zoom and center automatically if the props get changed
    componentWillReceiveProps(nextProps) {
        if (!lodash_1.isEqual(nextProps.center, this.props.center)) {
            this.setCenter(nextProps.center);
        }
        if (nextProps.zoom !== this.props.zoom) {
            this.setZoom(nextProps.zoom);
        }
    }
});
exports.default = exports.HMapMethods;
