// react-here-maps/dist/Circle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
// export the Circle React component from this module
class Circle extends React.Component {
    // change the position automatically if the props get changed
    componentWillReceiveProps(nextProps) {
        if (nextProps.lat !== this.props.lat || nextProps.lng !== this.props.lng) {
            this.setCenter({
                lat: nextProps.lat,
                lng: nextProps.lng
            });
        }
        if (nextProps.radius !== this.props.radius) {
            this.setRadius(nextProps.radius);
        }
    }
    // remove the circle on unmount of the component
    componentWillUnmount() {
        const { map } = this.context;
        if (this.circle) {
            map.removeObject(this.circle);
        }
    }
    render() {
        const { map } = this.context;
        if (map && !this.circle) {
            this.addCircleToMap();
        }
        return null;
    }
    addCircleToMap() {
        const { map, } = this.context;
        const { lat, lng, strokeColor, lineWidth, fillColor, radius, } = this.props;
        // create a circle at the provided location
        const circle = new H.map.Circle({
            lat,
            lng
        }, radius, {
            style: {
                fillColor,
                lineWidth,
                strokeColor
            }
        });
        map.addObject(circle);
        this.circle = circle;
    }
    setCenter(point) {
        this.circle.setCenter(point);
    }
    setRadius(radius) {
        this.circle.setRadius(radius);
    }
}
// define the context types that are passed down from a <HEREMap> instance
Circle.contextTypes = {
    map: React.PropTypes.object
};
Circle.defaultProps = {
    fillColor: "rgba(255, 255, 255, 0.5)",
    lineWidth: 1,
    radius: 1000,
    strokeColor: "black"
};
exports.Circle = Circle;
// make the Circle component the default export
exports.default = Circle;
